statistics/handle/cache/internal/lfu/lfu_cache_test.go

```diff
@@ -0,0 +1,165 @@
+	t4 := testutil.NewMockStatisticsTable(2, 1, true, false, false)
+	lfu.Put(int64(1), t4, false)
+	require.Equal(t, lfu.Cost(), 6*mockCMSMemoryUsage+7*mockCMSMemoryUsage)
+	t5 := testutil.NewMockStatisticsTable(2, 2, true, false, false)
+	lfu.Put(int64(1), t5, false)
+	require.Equal(t, lfu.Cost(), 7*mockCMSMemoryUsage+7*mockCMSMemoryUsage)
+
+	t6 := testutil.NewMockStatisticsTable(1, 2, true, false, false)
+	lfu.Put(int64(1), t6, false)
+	require.Equal(t, lfu.Cost(), 7*mockCMSMemoryUsage+6*mockCMSMemoryUsage)
+
+	t7 := testutil.NewMockStatisticsTable(1, 1, true, false, false)
+	lfu.Put(int64(1), t7, false)
+	require.Equal(t, lfu.Cost(), 6*mockCMSMemoryUsage+6*mockCMSMemoryUsage)
+	lfu.Wait()
+	require.Equal(t, uint64(lfu.Cost()), lfu.metrics().CostAdded()-lfu.metrics().CostEvicted())
+}
+
+func TestLFUPutTooBig(t *testing.T) {
+	lfu, err := NewLFU(1)
+	require.NoError(t, err)
+	mockTable := testutil.NewMockStatisticsTable(1, 1, true, false, false)
+	// put mockTable, the index should be evicted
+	lfu.Put(int64(1), mockTable, false)
+	_, ok := lfu.Get(int64(1), false)
+	require.False(t, ok)
+	lfu.Wait()
+	require.Equal(t, uint64(lfu.Cost()), lfu.metrics().CostAdded()-lfu.metrics().CostEvicted())
+}
+
+func TestCacheLen(t *testing.T) {
+	capacity := int64(12)
+	lfu, err := NewLFU(capacity)
+	require.NoError(t, err)
+	t1 := testutil.NewMockStatisticsTable(2, 1, true, false, false)
+	require.Equal(t, int64(12), t1.MemoryUsage().TotalTrackingMemUsage())
+	lfu.Put(int64(1), t1, false)
+	t2 := testutil.NewMockStatisticsTable(1, 1, true, false, false)
+	// put t2, t1 should be evicted 2 items and still exists in the list
+	lfu.Put(int64(2), t2, false)
+	lfu.Wait()
+	require.Equal(t, lfu.Len(), 1)
+	require.Equal(t, uint64(lfu.Cost()), lfu.metrics().CostAdded()-lfu.metrics().CostEvicted())
+
+	// put t3, t1/t2 should be evicted all items and disappeared from the list
+	t3 := testutil.NewMockStatisticsTable(2, 1, true, false, false)
+	lfu.Put(int64(3), t3, false)
+	lfu.Wait()
+	require.Equal(t, lfu.Len(), 1)
+	require.Equal(t, uint64(lfu.Cost()), lfu.metrics().CostAdded()-lfu.metrics().CostEvicted())
+}
+
+func TestLFUCachePutGetWithManyConcurrency(t *testing.T) {
+	// to test DATA RACE
+	capacity := int64(12)
+	lfu, err := NewLFU(capacity)
+	require.NoError(t, err)
+	var wg sync.WaitGroup
+	for i := 0; i < 1000; i++ {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			t1 := testutil.NewMockStatisticsTable(1, 1, true, false, false)
+			lfu.Put(int64(i), t1, true)
+		}(i)
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			lfu.Get(int64(i), true)
+		}(i)
+	}
+	wg.Wait()
+	lfu.Wait()
+	require.Equal(t, uint64(lfu.Cost()), lfu.metrics().CostAdded()-lfu.metrics().CostEvicted())
+}
+
+func TestLFUCachePutGetWithManyConcurrency2(t *testing.T) {
+	// to test DATA RACE
+	capacity := int64(12)
+	lfu, err := NewLFU(capacity)
+	require.NoError(t, err)
+	var wg sync.WaitGroup
+	var cnt atomic.Int64
```

Comment:
This variable seems useless... This variable is used to test for what?

Reply:
removed.